pallets/hsm/src/lib.rs

```diff
@@ -0,0 +1,1388 @@
+	///
+	/// This function handles:
+	/// 1. Calculating the purchase price with fees
+	/// 2. Ensuring maximum collateral holding isn't exceeded
+	/// 3. Transferring collateral from user to HSM
+	/// 4. Minting new Hollar to the user
+	///
+	/// Returns the Hollar and collateral amounts traded.
+	#[require_transactional]
+	fn do_trade_hollar_out(
+		who: &T::AccountId,
+		collateral_asset: T::AssetId,
+		calculate_amounts: impl FnOnce(Price) -> Result<(Balance, Balance), DispatchError>,
+	) -> Result<(Balance, Balance), DispatchError> {
+		let collateral_info = Collaterals::<T>::get(collateral_asset).ok_or(Error::<T>::AssetNotApproved)?;
+		let pool_state = Self::get_stablepool_state(collateral_info.pool_id)?;
+		let collateral_pos = pool_state
+			.asset_idx(collateral_asset)
+			.ok_or(Error::<T>::AssetNotFound)?;
+		let peg = pool_state.pegs[collateral_pos]; //hollar/collateral
+		let purchase_price = math::calculate_purchase_price(peg, collateral_info.purchase_fee);
+
+		let (hollar_amount, collateral_amount) = calculate_amounts(purchase_price)?;
+
+		ensure!(
+			Self::ensure_max_collateral_holding(collateral_asset, &collateral_info, collateral_amount),
+			Error::<T>::MaxHoldingExceeded
+		);
+
+		<T as Config>::Currency::transfer(
+			collateral_asset,
+			who,
+			&Self::account_id(),
+			collateral_amount,
+			Preservation::Expendable,
+		)?;
+
+		Self::mint_hollar(who, hollar_amount)?;
+
+		Ok((hollar_amount, collateral_amount))
+	}
+
+	/// Retrieve hollar contract address
+	fn get_hollar_contract_address() -> Result<EvmAddress, DispatchError> {
+		if cfg!(feature = "runtime-benchmarks") {
+			// for benchmarking purposes, we simply return some address
+			// it is because we dont have Hollar registered in registry as Erc20,
+			// but we still read the hollar from registry storage, to account for registry read weight
+			let _ = T::GhoContractAddress::contract_address(T::HollarId::get());
+			Ok(EvmAddress::from_slice(&hex!(
+				"0101010101010101010101010101010101010101"
+			)))
+		} else {
+			T::GhoContractAddress::contract_address(T::HollarId::get())
+				.ok_or(Error::<T>::HollarContractAddressNotFound.into())
+		}
+	}
+
+	/// Mint Hollar by calling the GHO token contract
+	///
+	/// Creates new Hollar tokens by interacting with the GHO ERC20 contract.
+	/// The HSM pallet acts as the facilitator for minting.
+	///
+	/// Returns Ok if successful, or an error if the EVM interaction fails.
+	fn mint_hollar(who: &T::AccountId, amount: Balance) -> DispatchResult {
+		let contract = Self::get_hollar_contract_address()?;
+		let pallet_address = T::EvmAccounts::evm_address(&Self::account_id());
+
+		// Create the context for the EVM call
+		let context = CallContext::new_call(contract, pallet_address);
+
+		// Encode the mint function call with recipient and amount
+		let recipient_evm = T::EvmAccounts::evm_address(who);
+		let mut data = Into::<u32>::into(ERC20Function::Mint).to_be_bytes().to_vec();
```

Comment:
use EvmDataWriter, it is way cleaner

Reply:
i wanted - but it is only avaiable in the runtime which i dont have access to it in a pallet.

Reply:
thats weird, it doesnt depend on anything there so it should be probably moved to some utility crate

Reply:
wdyt @dmoka ?